### PR TITLE
Brandon/better drivetrain autos

### PR DESCRIPTION
**Pull request by BrandonHarad**

### CODE REVIEW

File: src/main/java/frc/robot/Constants.java [hunk 1]
```diff
@@ -12,12 +16,21 @@
...
+
+ //TODO Find the correct value for all these constants
 public final class Constants {
+  public static final double KP_DRIVE_VEL = 3.2925;
+public static final double KS_VOLTS = 0.650;
+public static final double KV_VOLT_SECONDS_PER_METER = 2.81;
+public static final double KA_VOLT_SECONDS_SQUARED_PER_METER = 0.224;
+public static final double ENCODER_TICKS_PER_METER = 2048;
+// TODO: I never measured anything (I want to die), so make sure to replace it all :)))))))))
+  public static final double K_TRACKWIDTH_METERS = 0.6604;
```
Comment (autumnwilkes) [inline src/main/java/frc/robot/Constants.java:28]: Are these actually not real?

File: src/main/java/frc/robot/Constants.java [hunk 2]
```diff
@@ -26,22 +39,49 @@ public static class DrivetrainConstants {
...
 
+  public static class PneumaticsConstants {
+    public static final int CLAW_ID = 0;
 
+    public static final int SHOOTER_ID_1 = 1;
+    public static final int SHOOTER_ID_2 = 2;
+    public static final int SHOOTER_ID_3 = 3;
+    public static final int SHOOTER_ID_4 = 4;
+    public static final int SHOOTER_ID_5 = 5;
   }
```
Comment (autumnwilkes) [inline src/main/java/frc/robot/Constants.java]: Why is this existing?

File: src/main/java/frc/robot/Robot.java [hunk 3]
```diff
@@ -28,6 +36,13 @@ public void robotInit() {
     // Instantiate our RobotContainer.  This will perform all our button bindings, and put our
     // autonomous chooser on the dashboard.
     m_robotContainer = new RobotContainer();
+    m_chooser.setDefaultOption("Ram Auto", kDefaultAuto);
```
Comment (autumnwilkes) [inline src/main/java/frc/robot/Robot.java:38]: Seems potentially dangerous to have the default option be Ram Auto (maybe just me tho)

File: src/main/java/frc/robot/RobotContainer.java [hunk 4]
```diff
@@ -27,25 +31,26 @@
...
-      new Elevator();
+          new Elevator();
   private final PneumaticsControlModule m_pcm =
           new PneumaticsControlModule(OperatorConstants.kPCM_ID);
   private final Shooter m_shooter =
           new Shooter(m_pcm);
   private final Grabber m_grabber =
           new Grabber(m_pcm);
 
+  private final GrabberCommand m_grabberCommand = new GrabberCommand(m_grabber);
```
Comment (autumnwilkes) [inline src/main/java/frc/robot/RobotContainer.java]: Why is this here?

File: src/main/java/frc/robot/commands/Drivetrain/Drivetrain.java [hunk 5]
```diff
@@ -1,36 +1,93 @@
...
+        odometry.update(
+            pigeon.getRotation2d(), 
+            leftFront.getSelectedSensorPosition()/Constants.ENCODER_TICKS_PER_METER,
+            rightFront.getSelectedSensorPosition()/Constants.ENCODER_TICKS_PER_METER
+        );
+    }
+    
+    public void setPercent(double leftPercent, double rightPercent){
+        leftFront.set(ControlMode.PercentOutput, leftPercent);
+        rightFront.set(ControlMode.PercentOutput, -rightPercent);
```
Comment (autumnwilkes) [inline src/main/java/frc/robot/commands/Drivetrain/Drivetrain.java]: Why is this negative?

File: src/main/java/frc/robot/commands/Drivetrain/Drivetrain.java [hunk 6]
```diff
@@ -1,36 +1,93 @@
...
 
-        frontLeft.setInverted(true);
-        backLeft.setInverted(true);
-        frontRight.setInverted(false);
-        backRight.setInverted(false);
+    public TalonFXSensorCollection getLeftSensors(){
+        return leftFront.getSensorCollection();
+    }
+    public TalonFXSensorCollection getRightSensors(){
+        return leftFront.getSensorCollection();
```
Comment (autumnwilkes) [inline src/main/java/frc/robot/commands/Drivetrain/Drivetrain.java]: This says right yet gets left

File: src/main/java/frc/robot/commands/Drivetrain/AutoCommand.java
Comment (autumnwilkes) [inline src/main/java/frc/robot/commands/Drivetrain/AutoCommand.java:1]: This seems like a very not scalable way of writing autos. As a result of it, a lot of the variables are very confusingly named. Also it does not seem that there is actually a `RamAuto` being run, unless if you are running `RamAuto`, then expecting to be able to pick up bunnies

File: src/main/java/frc/robot/commands/Drivetrain/Drivetrain.java
Comment (autumnwilkes) [inline src/main/java/frc/robot/commands/Drivetrain/Drivetrain.java:1]: Wasn't this to be rewritten with Phoenix6?

File: src/main/java/frc/robot/commands/Drivetrain/AutoCommand.java [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+        switch (autoChoice) {
+        case "DoNothing":
+            drivetrain.resetOdometry(trajectory.getInitialPose());
+
+            addCommands(
+                    //      new InstantCommand(()-> SmartDashboard.putNumber("navx/endRotation", drivetrain.getPose().getRotation().getDegrees())), common Simon L
+                    new InstantCommand(()-> drivetrain.setPercent(0, 0))
+            );
+            break;
+        case "GrabBunny":
```
Comment (autumnwilkes) [inline src/main/java/frc/robot/commands/Drivetrain/AutoCommand.java:47]: Why are `RamAuto` and `UnramAuto` being used for GrabBunny?

File: src/main/deploy/pathplanner/GrabBunny.path
Comment (autumnwilkes) [inline src/main/deploy/pathplanner/GrabBunny.path:1]: This GrabBunny path is not used in any of the autos

File: src/main/java/frc/robot/Constants.java [hunk 8]
```diff
@@ -26,22 +39,49 @@ public static class DrivetrainConstants {
     public static final int FRONT_RIGHT_KEY = 4;
     public static final int BACK_LEFT_KEY = 1;
     public static final int BACK_RIGHT_KEY = 3;
+    public static final double ENCODER_TICKS_PER_METER = 2048;
```
Comment (rutmanz) [inline src/main/java/frc/robot/Constants.java]: This is encoder ticks per rotation, no?

File: src/main/java/frc/robot/Constants.java [hunk 9]
```diff
@@ -26,22 +39,49 @@ public static class DrivetrainConstants {
...
+  }
 
+  public static class PneumaticsConstants {
+    public static final int CLAW_ID = 0;
 
+    public static final int SHOOTER_ID_1 = 1;
+    public static final int SHOOTER_ID_2 = 2;
+    public static final int SHOOTER_ID_3 = 3;
+    public static final int SHOOTER_ID_4 = 4;
+    public static final int SHOOTER_ID_5 = 5;
```
Comment (rutmanz) [inline src/main/java/frc/robot/Constants.java]: What are these for?

File: src/main/java/frc/robot/Robot.java [hunk 10]
```diff
@@ -15,8 +18,13 @@
...
  */
 public class Robot extends TimedRobot {
-  private Command m_autonomousCommand;
-
+  private static final String kDefaultAuto = "RamAuto";
+  private static final String kCustomAuto1 = "DoNothing";
+  private static final String kCustomAuto2 = "InNOut";
+  private static final String kCustomAuto3 = "GrabBunny";
+  private static final String kCustomAuto4 = "SprayNPraySeq";
+  private static final String kCustomAuto5 = "SprayNPrayFull";
```
Comment (rutmanz) [inline src/main/java/frc/robot/Robot.java:26]: These constants aren't actually used in the switch case

File: src/main/java/frc/robot/commands/Drivetrain/AutoCommand.java [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+import frc.robot.commands.Grabber.Grabber;
+import frc.robot.commands.Grabber.GrabberCommand;
+import frc.robot.commands.Shooter.Shooter;
+import frc.robot.commands.Shooter.ShooterCommand;
+
+import java.nio.file.Path;
+
+public class AutoCommand extends SequentialCommandGroup{
+    public AutoCommand(Drivetrain drivetrain, String autoChoice, Elevator elevator, Grabber grabber, Shooter shooter){
+        addRequirements(drivetrain);
```
Comment (rutmanz) [inline src/main/java/frc/robot/commands/Drivetrain/AutoCommand.java:28]: Should also add elevator and shooter, don't want those manual controls to get in the way

File: src/main/java/frc/robot/commands/Drivetrain/AutoCommand.java [hunk 12]
```diff
@@ -0,0 +1,204 @@
...
+            );
+
+
+            drivetrain.resetOdometry(trajectory.getInitialPose());
+
+            addCommands(
+                ramCommand,
+                new ElevatorUp(elevator),
+                grabber.setTrue,
+                grabber.setFalse,
```
Comment (rutmanz) [inline src/main/java/frc/robot/commands/Drivetrain/AutoCommand.java]: What...

File: src/main/java/frc/robot/commands/Drivetrain/AutoCommand.java [hunk 13]
```diff
@@ -0,0 +1,204 @@
...
+                    drivetrain::setVolts,
+                    drivetrain
+            );
+
+            drivetrain.resetOdometry(trajectory.getInitialPose());
+            ShooterCommand sc0 = new ShooterCommand(shooter, 0);
+            ShooterCommand sc1 = new ShooterCommand(shooter, 1);
+            ShooterCommand sc2 = new ShooterCommand(shooter, 2);
+            ShooterCommand sc3 = new ShooterCommand(shooter, 3);
+            ShooterCommand sc4 = new ShooterCommand(shooter, 4);
```
Comment (rutmanz) [inline src/main/java/frc/robot/commands/Drivetrain/AutoCommand.java:134]: It's ok to declare these inside addCommands, they don't need to be done with extra variables

File: src/main/java/frc/robot/commands/Drivetrain/AutoCommand.java [hunk 14]
```diff
@@ -0,0 +1,204 @@
...
+
+            drivetrain.resetOdometry(trajectory.getInitialPose());
+            ShooterCommand sc01 = new ShooterCommand(shooter, 0);
+            ShooterCommand sc11 = new ShooterCommand(shooter, 1);
+            ShooterCommand sc21 = new ShooterCommand(shooter, 2);
+            ShooterCommand sc31 = new ShooterCommand(shooter, 3);
+            ShooterCommand sc41 = new ShooterCommand(shooter, 4);
+
+            addCommands(
+                    new ParallelCommandGroup(ramCommand5, sc01, sc21, sc31, sc41),
```
Comment (rutmanz) [inline src/main/java/frc/robot/commands/Drivetrain/AutoCommand.java:171]: We're just skipping sc11? Same comment as before on the inclining, these variable names are declining in readability

File: src/main/java/frc/robot/commands/Drivetrain/AutoCommand.java [hunk 15]
```diff
@@ -0,0 +1,204 @@
...
+                    new InstantCommand(() -> drivetrain.setPercent(0, 0))
+            );
+            break;
+        default:
+            trajectory = PathPlanner.loadPath("RamAuto", new PathConstraints(3.5, 2));
```
Comment (rutmanz) [inline src/main/java/frc/robot/commands/Drivetrain/AutoCommand.java:250]: Adley pointed this out, moving should not be the default auto. If they didn't select one, it should not move

File: src/main/java/frc/robot/commands/Drivetrain/AutoCommand.java [hunk 16]
```diff
@@ -0,0 +1,204 @@
...
+import frc.robot.commands.Elevator.ElevatorDown;
+import frc.robot.commands.Elevator.ElevatorUp;
+import frc.robot.commands.Grabber.Grabber;
+import frc.robot.commands.Grabber.GrabberCommand;
+import frc.robot.commands.Shooter.Shooter;
+import frc.robot.commands.Shooter.ShooterCommand;
+
+import java.nio.file.Path;
+
+public class AutoCommand extends SequentialCommandGroup{
```
Comment (rutmanz) [inline src/main/java/frc/robot/commands/Drivetrain/AutoCommand.java:26]: So this is definitely a way that you _can_ make multiple auto sequences, but why not make multiple Commands, one for each auto? They're not really reusing anything between all of them

File: src/main/java/frc/robot/commands/Drivetrain/DriveCommand.java [hunk 17]
```diff
@@ -41,9 +39,9 @@ public void execute() {
         rightInput *= 0.5;
 
         if (invertDrive) {
```
Comment (rutmanz) [inline src/main/java/frc/robot/commands/Drivetrain/DriveCommand.java:41]: I noticed this when driving in teleop yesterday, since you inverted one motor side, you need to invert one of the input sides relative to the other

File: src/main/java/frc/robot/commands/Drivetrain/Drivetrain.java [hunk 18]
```diff
@@ -1,36 +1,93 @@
...
+    }
+
+    public void setVolts(double leftVolts, double rightVolts){
+        leftFront.set(ControlMode.PercentOutput, leftVolts/12);
+        rightFront.set(ControlMode.PercentOutput, rightVolts/12);
+    }
+    public DifferentialDriveWheelSpeeds getWheelSpeeds(){
+        return new DifferentialDriveWheelSpeeds(
+                leftFront.getSelectedSensorVelocity()*10/Constants.ENCODER_TICKS_PER_METER,
+                rightFront.getSelectedSensorVelocity()*10/Constants.ENCODER_TICKS_PER_METER
```
Comment (rutmanz) [inline src/main/java/frc/robot/commands/Drivetrain/Drivetrain.java:63]: Why 10?

File: src/main/java/frc/robot/commands/Grabber/Grabber.java [hunk 19]
```diff
@@ -18,6 +20,7 @@ public boolean get() {
         return claw.get();
     }
 
-
+    public InstantCommand setTrue = new InstantCommand(() -> this.set(true));
+    public InstantCommand setFalse = new InstantCommand(() -> this.set(false));
```
Comment (rutmanz) [inline src/main/java/frc/robot/commands/Grabber/Grabber.java]: You don't want to have essentially constant command instances like this, because trying to add it to multiple groups will make the command scheduler angry. Instead you should probably make a factory method that just returns a new instant command of your preferred action

**Review by rutmanz — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/Constants.java [hunk 20]
```diff
@@ -26,22 +39,49 @@ public static class DrivetrainConstants {
...
     public static final int BACK_LEFT_KEY = 1;
     public static final int BACK_RIGHT_KEY = 3;
+    public static final double ENCODER_TICKS_PER_METER = 2048;
+
+
+    // to be clear, one of us needs to use SI to find these constants
+    // they are not all just one, goofily
+    public static final double ksVolts = 1;
+    public static final double kvVoltSecondsPerMeter = 1;
+    public static final double kaVoltSecondsSquaredPerMeter = 1;
```
Comment (autumnwilkes) [inline src/main/java/frc/robot/Constants.java]: These are already defined earlier in the Constants class with slightly different names

File: src/main/java/frc/robot/commands/Drivetrain/AutoCommand.java [hunk 21]
```diff
@@ -0,0 +1,204 @@
...
+        case "DoNothing":
+            drivetrain.resetOdometry(trajectory.getInitialPose());
+
+            addCommands(
+                    //      new InstantCommand(()-> SmartDashboard.putNumber("navx/endRotation", drivetrain.getPose().getRotation().getDegrees())), common Simon L
+                    new InstantCommand(()-> drivetrain.setPercent(0, 0))
+            );
+            break;
+        case "GrabBunny":
+            trajectory = PathPlanner.loadPath("RamAuto", new PathConstraints(3.5, 2));
```
Comment (autumnwilkes) [inline src/main/java/frc/robot/commands/Drivetrain/AutoCommand.java]: It appears that, rather than RamAuto being the auto that, well, rams, it is just an auto that goes out into the field?

File: src/main/java/frc/robot/commands/Drivetrain/AutoCommand.java [hunk 22]
```diff
@@ -0,0 +1,204 @@
...
+            PathPlannerTrajectory trajectory1 = PathPlanner.loadPath("UnRamAuto", new PathConstraints(3.5, 2));
+
+            PPRamseteCommand ramCommand = new PPRamseteCommand(
```
Comment (autumnwilkes) [inline src/main/java/frc/robot/commands/Drivetrain/AutoCommand.java:51]: Why are these being called ramCommands?

File: src/main/java/frc/robot/Constants.java (hunk 9 above)
Comment (BrandonHarad) [reply src/main/java/frc/robot/Constants.java]: Nothing

File: src/main/java/frc/robot/Robot.java (hunk 10 above)
Comment (BrandonHarad) [reply src/main/java/frc/robot/Robot.java:26]: ok

File: src/main/java/frc/robot/Constants.java (hunk 2 above)
Comment (BrandonHarad) [reply src/main/java/frc/robot/Constants.java]: *Why does this exist? And I don't know

File: src/main/java/frc/robot/commands/Drivetrain/Drivetrain.java (hunk 18 above)
Comment (BrandonHarad) [reply src/main/java/frc/robot/commands/Drivetrain/Drivetrain.java:63]: Used in Simon's code for Bread

File: src/main/java/frc/robot/commands/Drivetrain/AutoCommand.java
Comment (BrandonHarad) [reply src/main/java/frc/robot/commands/Drivetrain/AutoCommand.java:1]: I have no idea what you mean

File: src/main/java/frc/robot/commands/Drivetrain/Drivetrain.java
Comment (BrandonHarad) [reply src/main/java/frc/robot/commands/Drivetrain/Drivetrain.java:1]: It was, but we held off on updating the motors due to time pressure.

File: src/main/java/frc/robot/commands/Drivetrain/AutoCommand.java (hunk 16 above)
Comment (BrandonHarad) [reply src/main/java/frc/robot/commands/Drivetrain/AutoCommand.java:26]: That seems reasonable. I can't be bothered tho
